### PR TITLE
Fix/staking pool

### PR DESCRIPTION
**Pull request by Akkii4**

### CODE REVIEW

**Review by saurabh-lodha-16 — APPROVED** [review]

File: contracts/RewardManagerFactory.sol [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    using SafeMath for uint128;
+
+    /// @notice all the information for this RewardManager in one struct
+    struct RewardManagerInfo {
+        address managerAddress;
+        uint256 startDistribution;
+        uint256 endDistribution;
+    }
+    
+    struct tempInfo{
```
Comment (ashwinYardi) [inline contracts/RewardManagerFactory.sol:17]: UserInfo

File: contracts/RewardManagerFactory.sol [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        uint256 _totalDrawnAmount;
+        uint256 _amountBurnt;
+        uint256 _claimable;
+        uint256 _bonusRewards;
+        uint256 _stillDue;
+    }
+
+    RewardManagerInfo[] public managers;
+
+    uint256 public managerIndex;
```
Comment (ashwinYardi) [inline contracts/RewardManagerFactory.sol:28]: Why is this required exactly?

File: contracts/RewardManagerFactory.sol [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    
+    function removeRewardManager(uint256 _index) public onlyOwner {
+        require(_index <= managerIndex, "Invalid Index");
+        delete managers[_index];
+    }
+    
+    function userTotalVestingInfo(address _user)
+    public view
+    returns (uint256 totalVested, uint256 totalDrawnAmount, uint256 amountBurnt, uint256 claimable, uint256 bonusRewards, uint256 stillDue){
+        tempInfo memory temp;
```
Comment (ashwinYardi) [inline contracts/RewardManagerFactory.sol:109]: UserInfo and User

File: contracts/RewardManagerFactory.sol [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        for(uint256 i = 0; i < totalRewardManagers; i++){
+            address rewardManagerAddress = managers[i].managerAddress;
+            if(rewardManagerAddress != address(0)){
+                RewardManager manager = RewardManager(rewardManagerAddress);
+                (temp._totalVested, temp._totalDrawnAmount, temp._amountBurnt, temp._claimable, temp._bonusRewards, temp._stillDue) = manager.vestingInfo(_user);
```
Comment (ashwinYardi) [inline contracts/RewardManagerFactory.sol:114]: Lets check here if there is non zero amount pending to be claimed. Saving few addition operations for alrady claimed reward managers.

**Review by ashwinYardi — COMMENTED** [review]
Can we also please use the linter and format the contracts?

File: contracts/RewardManagerFactory.sol (hunk 2 above)
Comment (Akkii4) [reply contracts/RewardManagerFactory.sol]: this was being removed in previous commit

File: contracts/RewardManagerFactory.sol [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+                (
+                    user._totalVested,
+                    user._totalDrawnAmount,
+                    user._amountBurnt,
+                    user._claimable,
+                    user._bonusRewards,
+                    user._stillDue
+                ) = manager.vestingInfo(_user);
+                
+                if(user._totalVested) {
```
Comment (ashwinYardi) [inline contracts/RewardManagerFactory.sol:126]: `if ( _totalVested.sub(_totalDrawnAmount) > 0 )`

File: contracts/RewardManagerFactory.sol [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+    /**
+     * @notice Draws down any vested tokens due in all Reward Manager
+     * @dev Must be called directly by the beneficiary assigned the tokens in the vesting
+     */
+    function drawDown() external onlyOwner {
+        for (uint256 i = 0; i < totalRewardManagers; i++) {
+            address rewardManagerAddress = managers[i].managerAddress;
+            if (rewardManagerAddress != address(0)) {
+                RewardManager manager = RewardManager(rewardManagerAddress);
```
Comment (ashwinYardi) [inline contracts/RewardManagerFactory.sol:146]: `if ( _totalVested.sub(_totalDrawnAmount) > 0  && startDistribution < block.timestamp)`

File: contracts/RewardManagerFactory.sol [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+    /**
+     * @notice Pre maturely Draws down all vested tokens by burning the preMaturePenalty
+     * @dev Must be called directly by the beneficiary assigned the tokens in the vesting
+     */
+    function preMatureDraw() external onlyOwner {
+        for (uint256 i = 0; i < totalRewardManagers; i++) {
+            address rewardManagerAddress = managers[i].managerAddress;
+            if (rewardManagerAddress != address(0)) {
+                RewardManager manager = RewardManager(rewardManagerAddress);
```
Comment (ashwinYardi) [inline contracts/RewardManagerFactory.sol:160]: `if ( _totalVested.sub(_totalDrawnAmount) > 0`

File: contracts/RewardManagerFactory.sol [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+                manager.preMatureDraw(msg.sender);
+            }
+        }
+    }
+
+    function updatePreMaturePenalty(
+        uint256 _index,
```
Comment (ashwinYardi) [inline contracts/RewardManagerFactory.sol:167]: Lets add `address -> index` mapping as well? So that we can check easily whats the index we need to use.

**Review by ashwinYardi — CHANGES_REQUESTED** [review]

File: contracts/RewardManagerFactory.sol (hunk 7 above)
Comment (Akkii4) [reply contracts/RewardManagerFactory.sol:160]: but this logic is being already handled in preMatureDraw in RewardManager & any user who's remaining Balance equals to 0 will receive error while triggering this function in factory

File: contracts/RewardManagerFactory.sol (hunk 5 above)
Comment (Akkii4) [reply contracts/RewardManagerFactory.sol:126]: but as this is a view function & shows user total vesting info , so despite a user has pending claimable balance or not , shouldn't that rewardManger data be used for getting his overall data

File: contracts/RewardManagerFactory.sol (hunk 6 above)
Comment (Akkii4) [reply contracts/RewardManagerFactory.sol:146]: this logic is being already handled in drawDown in RewardManager

Comment (ashwinYardi) [reply contracts/RewardManagerFactory.sol:146]: Its a require, it will revert resulting in function becoming unusable because at any point in time, at least one reward manager will be in accumulation phase which drawdown function will revert.

File: contracts/RewardManagerFactory.sol (hunk 5 above)
Comment (ashwinYardi) [reply contracts/RewardManagerFactory.sol:126]: Yes, correct, we should query and add all reward managers here.

File: contracts/RewardManagerFactory.sol (hunk 6 above)
Comment (Akkii4) [reply contracts/RewardManagerFactory.sol:146]: oh okay , got it , thanks

**Review by ashwinYardi — APPROVED** [review]
